lib/compress/zstdmt_compress.c

```diff
@@ -993,134 +1001,162 @@ size_t ZSTDMT_initCStream(ZSTDMT_CCtx* zcs, int compressionLevel) {
+        if (ZSTD_isError(job.cSize)) {
+            DEBUGLOG(5, "ZSTDMT_flushProduced: job %u : compression error detected : %s",
+                        zcs->doneJobID, ZSTD_getErrorName(job.cSize));
+            ZSTDMT_waitForAllJobsCompleted(zcs);
+            ZSTDMT_releaseAllJobResources(zcs);
+            return job.cSize;
+        }
+        /* add frame checksum if necessary (can only happen once) */
+        if ( job.jobCompleted
+          && job.frameChecksumNeeded ) {
+            U32 const checksum = (U32)XXH64_digest(&zcs->xxhState);
+            DEBUGLOG(4, "ZSTDMT_flushProduced: writing checksum : %08X \n", checksum);
+            MEM_writeLE32((char*)job.dstBuff.start + job.cSize, checksum);
+            job.cSize += 4;
+            zcs->jobs[wJobID].cSize += 4;
+            zcs->jobs[wJobID].frameChecksumNeeded = 0;
         }
-        {   size_t const toWrite = MIN(job.cSize - job.dstFlushed, output->size - output->pos);
-            DEBUGLOG(5, "ZSTDMT_flushNextJob: Flushing %u bytes from job %u ", (U32)toWrite, zcs->doneJobID);
+        assert(job.cSize >= job.dstFlushed);
+        if (job.dstBuff.start != NULL) {  /* one buffer present : some job is ongoing */
+            size_t const toWrite = MIN(job.cSize - job.dstFlushed, output->size - output->pos);
+            DEBUGLOG(5, "ZSTDMT_flushProduced: Flushing %u bytes from job %u (completion:%.1f%%)",
+                        (U32)toWrite, zcs->doneJobID, (double)job.consumed / job.srcSize * 100);
             memcpy((char*)output->dst + output->pos, (const char*)job.dstBuff.start + job.dstFlushed, toWrite);
             output->pos += toWrite;
             job.dstFlushed += toWrite;
-        }
-        if (job.dstFlushed == job.cSize) {   /* output buffer fully flushed => move to next one */
-            ZSTDMT_releaseBuffer(zcs->bufPool, job.dstBuff);
-            zcs->jobs[wJobID].dstBuff = g_nullBuffer;
-            zcs->jobs[wJobID].jobCompleted = 0;
-            zcs->doneJobID++;
-            zcs->consumed += job.srcSize;
-            zcs->produced += job.cSize;
-        } else {
-            zcs->jobs[wJobID].dstFlushed = job.dstFlushed;
-        }
-        /* return value : how many bytes left in buffer ; fake it to 1 if unknown but >0 */
+
+            if ( job.jobCompleted
+              && (job.dstFlushed == job.cSize) ) {   /* output buffer fully flushed => move to next one */
+                DEBUGLOG(5, "Job %u completed (%u bytes), moving to next one",
+                        zcs->doneJobID, (U32)job.dstFlushed);
+                ZSTDMT_releaseBuffer(zcs->bufPool, job.dstBuff);
+                zcs->jobs[wJobID].dstBuff = g_nullBuffer;
+                zcs->jobs[wJobID].jobCompleted = 0;
+                zcs->consumed += job.srcSize;
+                zcs->produced += job.cSize;
+                zcs->doneJobID++;
+            } else {
+                zcs->jobs[wJobID].dstFlushed = job.dstFlushed;   /* remember how much was flushed for next attempt */
```

Comment:
Since this is updated without a lock, and with `job.jobCompleted == 0`, we have to be careful to never read/write `dstFlushed` in the worker thread.

Reply:
`dstFlushed` is indeed never used / read in the worker thread.
Maybe I should refactor `ZSTDMT_jobDescription` to clearly label which member is shared or not.